src/mds/CDir.cc

```diff
@@ -1620,178 +1791,30 @@ void CDir::_omap_fetched(bufferlist& hdrbl, map<string, bufferlist>& omap,
-	}
-	continue;
-      }
-
-      bool undef_inode = false;
-      if (dn) {
-	CInode *in = dn->get_linkage()->get_inode();
-	if (in) {
-	  dout(12) << "_fetched  had dentry " << *dn << dendl;
-	  if (in->state_test(CInode::STATE_REJOINUNDEF)) {
-	    undef_inodes.push_back(in);
-	    undef_inode = true;
-	  }
-	} else
-	  dout(12) << "_fetched  had NEG dentry " << *dn << dendl;
-      }
 
-      if (!dn || undef_inode) {
-	// add inode
-	CInode *in = cache->get_inode(inode_data.inode.ino, last);
-	if (!in || undef_inode) {
-	  if (undef_inode && in)
-	    in->first = first;
-	  else
-	    in = new CInode(cache, true, first, last);
-	  
-	  in->inode = inode_data.inode;
-	  // symlink?
-	  if (in->is_symlink()) 
-	    in->symlink = inode_data.symlink;
-	  
-	  in->dirfragtree.swap(inode_data.dirfragtree);
-	  in->xattrs.swap(inode_data.xattrs);
-	  in->old_inodes.swap(inode_data.old_inodes);
-	  in->oldest_snap = inode_data.oldest_snap;
-	  in->decode_snap_blob(inode_data.snap_blob);
-	  if (snaps && !in->snaprealm)
-	    in->purge_stale_snap_data(*snaps);
-
-	  if (!undef_inode) {
-	    cache->add_inode(in); // add
-	    dn = add_primary_dentry(dname, in, first, last); // link
-	  }
-	  dout(12) << "_fetched  got " << *dn << " " << *in << dendl;
-
-	  if (in->inode.is_dirty_rstat())
-	    in->mark_dirty_rstat();
-
-	  if (stray) {
-	    dn->state_set(CDentry::STATE_STRAY);
-	    if (in->inode.nlink == 0)
-	      in->state_set(CInode::STATE_ORPHAN);
-	  }
-
-	  //in->hack_accessed = false;
-	  //in->hack_load_stamp = ceph_clock_now(g_ceph_context);
-	  //num_new_inodes_loaded++;
-	} else {
-	  dout(0) << "_fetched  badness: got (but i already had) " << *in
-		  << " mode " << in->inode.mode
-		  << " mtime " << in->inode.mtime << dendl;
-	  string dirpath, inopath;
-	  this->inode->make_path_string(dirpath);
-	  in->make_path_string(inopath);
-	  clog->error() << "loaded dup inode " << inode_data.inode.ino
-	    << " [" << first << "," << last << "] v" << inode_data.inode.version
-	    << " at " << dirpath << "/" << dname
-	    << ", but inode " << in->vino() << " v" << in->inode.version
-	    << " already exists at " << inopath << "\n";
-	  continue;
-	}
-      }
-    } else {
-      dout(1) << "corrupt directory, i got tag char '" << type << "' pos "
-        << pos << dendl;
-      cache->mds->clog->error() << "Corrupt directory entry '" << p->first
-        << "' in dirfrag " << *this;
-      // TODO: add a mechanism for selectively marking a path
-      // damaged, rather than marking the whole rank damaged.
-      cache->mds->damaged();
-      assert(0);  // Unreachable: damaged() respawns us
+    CDentry *dn = NULL;
+    try {
+      dn = _load_dentry(
+          p->first, dname, last, p->second, pos, snaps,
+          &force_dirty, &undef_inodes);
+    } catch (const buffer::error &err) {
+#if 0
+      derr << "Error decoding dentry '" << p->first << "' in dirfrag "
+           << dirfrag() << ": " << err << dendl;
+#endif
```

Comment:
Indeed: when I commit an #if 0's it's never on purpose, just something I did momentarily and then later missed when tidying.